pkg/storage/ingest/version_test.go

```diff
@@ -44,3 +47,203 @@ func TestRecordVersionHeader(t *testing.T) {
+		err = DeserializeRecordContent(v2bytes, &wr, 2)
+		require.NoError(t, err)
+		require.Len(t, wr.Timeseries, 1)
+		expLabels := []mimirpb.LabelAdapter{{Name: "__name__", Value: "test_metric_total"}, {Name: "job", Value: "test_job"}}
+		require.Equal(t, expLabels, wr.Timeseries[0].Labels)
+		expSamples := []mimirpb.Sample{{Value: 123.456, TimestampMs: 1234567890}}
+		require.Equal(t, expSamples, wr.Timeseries[0].Samples)
+		expExemplars := []mimirpb.Exemplar{
+			{
+				Value:       123.456,
+				TimestampMs: 1234567890,
+				Labels:      []mimirpb.LabelAdapter{{Name: "__name__", Value: "test_metric_total"}, {Name: "traceID", Value: "1234567890abcdef"}}},
+		}
+		require.Equal(t, expExemplars, wr.Timeseries[0].Exemplars)
+		expHistograms := []mimirpb.Histogram{
+			{
+				Timestamp:      1234567890,
+				Count:          &mimirpb.Histogram_CountInt{CountInt: 10},
+				Sum:            100,
+				Schema:         3,
+				ZeroCount:      &mimirpb.Histogram_ZeroCountInt{ZeroCountInt: 0},
+				PositiveSpans:  []mimirpb.BucketSpan{{Offset: 0, Length: 1}},
+				PositiveDeltas: []int64{1},
+				NegativeSpans:  []mimirpb.BucketSpan{{Offset: 0, Length: 1}},
+				NegativeDeltas: []int64{1},
+			},
+		}
+		require.Equal(t, expHistograms, wr.Timeseries[0].Histograms)
+		expMetadata := []*mimirpb.MetricMetadata{{
+			Type:             mimirpb.COUNTER,
+			MetricFamilyName: "test_metric_total",
+			Help:             "Help for test_metric_total",
+			Unit:             "seconds",
+		}}
+		require.Equal(t, expMetadata, wr.Metadata)
+	})
+}
+
+func BenchmarkDeserializeRecordContent(b *testing.B) {
+	// Generate a write request in each version
+	reqv1 := &mimirpb.PreallocWriteRequest{
+		WriteRequest: mimirpb.WriteRequest{
+			Timeseries: make([]mimirpb.PreallocTimeseries, 10000),
+		},
+	}
+	for i := range reqv1.Timeseries {
+		reqv1.Timeseries[i] = mockPreallocTimeseries(fmt.Sprintf("series_%d", i))
+	}
+	v1bytes, err := reqv1.Marshal()
+	require.NoError(b, err)
+
+	reqv2 := &mimirpb.PreallocWriteRequest{
+		WriteRequest: mimirpb.WriteRequest{
+			TimeseriesRW2: make([]mimirpb.TimeSeriesRW2, 10000),
+			SymbolsRW2:    make([]string, 0, 1+1+10000),
+		},
+	}
+	reqv2.SymbolsRW2 = append(reqv2.SymbolsRW2, "")
+	reqv2.SymbolsRW2 = append(reqv2.SymbolsRW2, "__name__")
+	for i := range reqv2.TimeseriesRW2 {
+		reqv2.TimeseriesRW2[i] = mimirpb.TimeSeriesRW2{
+			LabelsRefs: []uint32{V2RecordSymbolOffset + 1, V2RecordSymbolOffset + uint32(i) + 2},
+			Samples:    []mimirpb.Sample{{TimestampMs: 1, Value: 2}},
+			Exemplars:  []mimirpb.ExemplarRW2{},
+		}
+		reqv2.SymbolsRW2 = append(reqv2.SymbolsRW2, fmt.Sprintf("series_%d", i))
+	}
+	v2bytes, err := reqv2.Marshal()
+	require.NoError(b, err)
+
+	b.Run("deserialize v1", func(b *testing.B) {
+		for range b.N {
+			wr := &mimirpb.PreallocWriteRequest{}
+			err := DeserializeRecordContent(v1bytes, wr, 1)
+			if err != nil {
+				b.Fatal(err)
+			}
+			defer mimirpb.ReuseSlice(wr.Timeseries)
```

Comment:
What's the value of this defer?

_Same comment applies below._

Reply:
Nothing! copy paste :smile: